From 8862a4f5026b989fd23e72d643f5405589b30223 Mon Sep 17 00:00:00 2001
From: Daniel Hsu <Daniel-Hsu@quantatw.com>
Date: Fri, 22 May 2026 14:04:41 +0800
Subject: [PATCH 4/4] cpld: lattice: Send program done command after Lattice
 XO5 update

Implement and invoke programDone() using the 0x5E command code in
finishUpdate() to properly finalize the CPLD firmware update sequence.

The corresponding change is under review in Gerrit:
https://gerrit.openbmc.org/c/openbmc/phosphor-bmc-code-mgmt/+/90473

Change-Id: I0a7ebbc4a9f685d5261758b3b04a0bbd682e7327
Signed-off-by: Daniel Hsu <Daniel-Hsu@quantatw.com>
---
 cpld/lattice/lattice_xo5_cpld.cpp | 26 ++++++++++++++++++++++++++
 cpld/lattice/lattice_xo5_cpld.hpp |  1 +
 2 files changed, 27 insertions(+)

diff --git a/cpld/lattice/lattice_xo5_cpld.cpp b/cpld/lattice/lattice_xo5_cpld.cpp
index f262758..906e8a0 100644
--- a/cpld/lattice/lattice_xo5_cpld.cpp
+++ b/cpld/lattice/lattice_xo5_cpld.cpp
@@ -13,6 +13,7 @@ enum class xo5Cmd : uint8_t
     sectorErase = 0xd8,
     pageProgram = 0x02,
     pageRead = 0x0b,
+    programDone = 0x5e,
     readUsercode = 0xc0
 };
 
@@ -315,6 +316,24 @@ sdbusplus::async::task<bool> LatticeXO5CPLD::readUserCode(uint32_t& userCode)
     co_return true;
 }
 
+sdbusplus::async::task<bool> LatticeXO5CPLD::programDone()
+{
+    std::vector<uint8_t> request = {};
+    std::vector<uint8_t> response = {};
+    request.push_back(static_cast<uint8_t>(xo5Cmd::programDone));
+    request.push_back(0x0);
+    request.push_back(0x0);
+    request.push_back(0x0);
+
+    if (!i2cInterface.sendReceive(request, response))
+    {
+        lg2::error("Failed to send program done request.");
+        co_return false;
+    }
+
+    co_return true;
+}
+
 sdbusplus::async::task<bool> LatticeXO5CPLD::prepareUpdate(const uint8_t* image,
                                                            size_t imageSize)
 {
@@ -376,6 +395,13 @@ sdbusplus::async::task<bool> LatticeXO5CPLD::finishUpdate()
         lg2::error("Verify cfg data failed.");
         co_return false;
     }
+
+    if (!(co_await programDone()))
+    {
+        lg2::error("Send program done request failed.");
+        co_return false;
+    }
+
     co_return true;
 }
 
diff --git a/cpld/lattice/lattice_xo5_cpld.hpp b/cpld/lattice/lattice_xo5_cpld.hpp
index 1ac84cc..16b6c00 100644
--- a/cpld/lattice/lattice_xo5_cpld.hpp
+++ b/cpld/lattice/lattice_xo5_cpld.hpp
@@ -33,6 +33,7 @@ class LatticeXO5CPLD : public LatticeBaseCPLD
     sdbusplus::async::task<bool> programPage(uint8_t block, uint8_t page,
                                              const std::vector<uint8_t>& data);
     sdbusplus::async::task<bool> verifyCfg();
+    sdbusplus::async::task<bool> programDone();
     sdbusplus::async::task<bool> readPage(uint8_t block, uint8_t page,
                                           std::vector<uint8_t>& data);
 };
-- 
2.53.0-Meta

